internal/webhook/external/github.go

```diff
@@ -0,0 +1,130 @@
+package external
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	gh "github.com/google/go-github/v71/github"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xhttp "github.com/akuity/kargo/internal/http"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+// githubHandler handles push events for github.
+// After the request has been authenticated,
+// the kubeclient is queried for all warehouses that contain a subscription
+// to the repo in question. Those warehouses are then patched with a special
+// annotation that signals down stream logic to refresh the warehouse.
+func githubHandler(c client.Client, token string) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
+		eventType := r.Header.Get("X-GitHub-Event")
+		if eventType != "push" {
```

Comment:
Perhaps leave a todo here about this being where we will eventually switch on event type to perform different actions (e.g. refresh Promotion on PR merge).